File: test/RegistrationProcessTest.java
import com.google.appengine.api.datastore.DatastoreService;
import com.google.appengine.api.datastore.DatastoreServiceFactory;
import com.google.appengine.api.datastore.Query;
import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
import java.util.HashMap;
import java.util.Map;
import models.User;
import org.junit.*;
import play.mvc.Http;
import play.test.*;
import play.mvc.Http.*;
import utils.TestXMPPService;
import utils.XMPPSend;

/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

/**
 *
 * @author lstoll
 */
public class RegistrationProcessTest extends FunctionalTest {
    
    private final LocalServiceTestHelper helper =
        new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());

    @Before
    public void setUp() {
        helper.setUp();
        XMPPSend.testSetUp();
    }

    @After
    public void tearDown() {
        helper.tearDown();
        XMPPSend.testTearDown();
    }

    @Test
    public void testThatIndexPageWorks() {
        Response response = GET("/");
        assertIsOk(response);
        assertContentType("text/html", response);
        assertCharset("utf-8", response);
    }

    // Test that a user submitting a email address results in an XMPP invitiation
    @Test
    public void testJidSubmissionInviteAndCreate() {
        Map files = new HashMap();
        Map<String, String> params = new HashMap();
        params.put("jid", "l@lds.li");
        Response resp = POST("/Register/submitJid", params, files);

        assertTrue(TestXMPPService.lastRequestInvitation);

        DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
        assertEquals(1, ds.prepare(new Query("User")).countEntities());
    }

    // Test that a user gets a welcome XMPP page with a link to the oauth page
    @Test
    public void testNewUserWelcomeLinkXMPP() {
        User u = new User();
        u.setXmppID("l@lds.li");
        u.insert();

        GET("/Cron/oauthCheck");

        assertEquals("l@lds.li", TestXMPPService.lastRequestJID);
        // TODO - update with the expected ID!
        assertTrue(TestXMPPService.lastMessage.getBody().contains("Registration/oauth"));
    }

    // Test that when OAUTH returns, user is sent to the usage page
}

File: test/RegistrationProcessTest.scala
import play.test._

import org.junit._
import org.scalatest.junit._
import org.scalatest._
import org.scalatest.matchers._
import com.google.appengine.tools.development.testing._
import com.google.appengine.api.datastore._
import java.util.{HashMap, Map}
import play.mvc.Http._
import scala.collection.JavaConversions._

import models._
import utils.TestXMPPService
import utils.XMPPSend

class RegistrationProcessTest extends FunctionalTest with ShouldMatchersForJUnit
                                                     with Browser
                                                     with play.test.Matchers {
  var helper = new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig)

  var playGAEDevEnv:com.google.apphosting.api.ApiProxy.Environment  = null
  
  @Before def setUp {
    helper.setUp
    //play.modules.gae.GAEPlugin.
    play.Play.plugins.foreach { p =>
      if (p.isInstanceOf[play.modules.gae.GAEPlugin]) {
        var pin = p.asInstanceOf[play.modules.gae.GAEPlugin]
        playGAEDevEnv = pin.devEnvironment
        pin.devEnvironment = null
      }
    }
    XMPPSend.testSetUp

  }
  
  @After def tearDown {
    helper.tearDown
    play.Play.plugins.foreach { p =>
      if (p.isInstanceOf[play.modules.gae.GAEPlugin]) {
        var pin = p.asInstanceOf[play.modules.gae.GAEPlugin]
        pin.devEnvironment = playGAEDevEnv
      }
    }
    XMPPSend.testTearDown
  }
  
  @Test def testThatIndexPageWorks {
    var response = GET("/")
    response.shouldBeOk
    response contentTypeShouldBe "text/html"
    response charsetShouldBe "utf-8"
  }
  
  // Test that a user submitting a email address results in an XMPP invitiation
  @Test def testJidSubmissionInviteAndCreate {
    var ds = DatastoreServiceFactory.getDatastoreService
    ds.prepare(new Query("User")).countEntities should equal (0)
    
    var files:Map[String, java.io.File] = new HashMap[String, java.io.File]
    var params:Map[String, String] = new HashMap[String, String]
    params.put("jid", "l@lds.li")
    var resp = POST("/Register/submitJid", params, files)
    
    TestXMPPService.lastRequestInvitation should be (true)
    
    ds.prepare(new Query("User")).countEntities should equal (1)
    TestXMPPService.reset
  }
  
  // Test that a user gets a welcome XMPP page with a link to the oauth page
  @Test def testNewUserWelcomeLinkXMPP {
    var ds = DatastoreServiceFactory.getDatastoreService
    ds.prepare(new Query("User")).countEntities should equal (0)
    
    var u = new User
    u.xmppID = "l@lds.li"
    u.save

    User.query.needsReminder.countAll should equal(1)

    var response = GET("/Cron/oauthCheck")
    response shouldBeOk
    
    ds.prepare(new Query("User")).countEntities should equal (1)
    
    TestXMPPService.lastRequestJID should equal ("l@lds.li")
    TestXMPPService.lastMessage should not equal (null)
    // TODO - match URL
    TestXMPPService.lastMessage.getBody().contains("twitter") should be (true)
  }
  
  // Test that when OAUTH returns, user is sent to the usage page
}
